### PR TITLE
[fix](function) money format

### PR DESCRIPTION
**Pull request by zhiqiang-hhhh**
money_format can not handle negative decimal correctly. Because method `do_money_format` did not process the negative frac_value in a right way.

Fix the problem by:

1. minor refactor to money format
2. method `do_money_format` will consider negative value, add append `-` manually
3. make sure money_format works well on decimalV2, decimalV3, float, double, integer

### CODE REVIEW

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run cloud_p0

**Comment by zhiqiang-hhhh** [comment]
run buildall

File: be/src/vec/functions/function_string.h [hunk 1]
```diff
@@ -2961,18 +2965,77 @@ class FunctionMoneyFormat : public IFunction {
...
+    }
+
+    // for DecimalV2, it has already been truncated, so we need this flag
+    if (do_truncate) {
+        // magic number 2: since we need to truncate frac_part to 2 digits
+        if (scale > 2) {
+            DCHECK(scale <= 38);
+            // do truncate
+            auto multiplier = common::exp10_i128(std::abs(static_cast<int>(scale - 2)));
+            frac_value = std::abs(frac_value / multiplier);
```
Comment (jacktengg) [inline be/src/vec/functions/function_string.h:3013]: 这里小数部分四舍五入是不是给弄没了呀？

Comment (zhiqiang-hhhh) [reply be/src/vec/functions/function_string.h:3013]: 我们不做四舍五入，做 truncate

**Review by jacktengg — DISMISSED** [review]

File: be/src/vec/functions/function_string.h [hunk 2]
```diff
@@ -3058,70 +3121,55 @@ struct MoneyFormatDecimalImpl {
...
         } else if (auto* decimal32_column =
                            check_and_get_column<ColumnDecimal<Decimal32>>(*col_ptr)) {
             const UInt32 scale = decimal32_column->get_scale();
-            // scale is up to 9, so exp10_i32 is enough
-            const auto multiplier = common::exp10_i32(std::abs(static_cast<int>(scale - 2)));
             for (size_t i = 0; i < input_rows_count; i++) {
-                Decimal32 frac_part = decimal32_column->get_fractional_part(i);
-                if (scale > 2) {
-                    int delta = ((frac_part % multiplier) << 1) > multiplier;
-                    frac_part = frac_part / multiplier + delta;
```
Comment (jacktengg) [inline be/src/vec/functions/function_string.h:3080]: Can fix the rounding problem:
```
                auto int_part = decimal32_column->get_whole_part(i);
                Decimal32 frac_part = decimal32_column->get_fractional_part(i);
                auto dec_value = decimal32_column->get_element(i);
                if (scale > 2) {
                    dec_value = (dec_value + multiplier / 2) / multiplier;
                    int_part = dec_value / 100;
                    frac_part = dec_value % 100;
                } else if (scale < 2) {
                    frac_part = frac_part * multiplier;
                }
```

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Comment by zhiqiang-hhhh** [comment]
run buildall

**Review by jacktengg — APPROVED** [review]
LGTM

**Review by HappenLee — APPROVED** [review]
LGTM

**Comment by HappenLee** [comment]
run performance

**Pull request merged**